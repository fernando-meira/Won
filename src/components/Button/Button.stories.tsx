import { Meta, Story } from '@storybook/react/types-6-0';

import Button from '.';

export default {
  title: 'Button',
  component: Button,
  argsTypes: {
    children: {
      types: 'string',
    },
  },
} as Meta;

export const Basic: Story = (args) => <Button {...args} />;

Basic.args = {
  children: 'Buy now',
};
