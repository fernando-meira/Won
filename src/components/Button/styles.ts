import styled, { css, DefaultTheme } from 'styled-components';

import { ButtonProps } from '.';

type WrapperProps = {
  hasIcon: boolean;
} & Pick<ButtonProps, 'size' | 'isFullWidth'>;

const containerModifiers = {
  isFullWidth: () => css`
    width: 100%;
  `,
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
};

export const Container = styled.button<WrapperProps>`
  ${({ theme, size, isFullWidth, hasIcon }) => css`
    border: none;
    padding: ${theme.spacings.xxsmall};
    border-radius: ${theme.border.radius};

    align-items: center;
    display: inline-flex;
    justify-content: center;

    cursor: pointer;
    text-decoration: none;
    color: ${theme.colors.white};
    background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);

    &:hover {
      background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
    }

    ${!!size && containerModifiers[size](theme)};
    ${hasIcon && containerModifiers.withIcon(theme)};
    ${isFullWidth && containerModifiers.isFullWidth()};
  `}
`;
