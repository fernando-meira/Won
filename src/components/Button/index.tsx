import React from 'react';
import * as S from './styles';

export type ButtonProps = {
  isFullWidth?: boolean;
  icon?: React.ReactNode;
  children?: React.ReactNode;
  size?: 'small' | 'medium' | 'large';
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void;
};

export default function Button({
  icon,
  children,
  size = 'medium',
  isFullWidth = false,
  ...rest
}: ButtonProps) {
  return (
    <S.Container
      {...rest}
      size={size}
      hasIcon={!!icon}
      isFullWidth={isFullWidth}
    >
      {!!icon && icon}

      {!!children && <span>{children}</span>}
    </S.Container>
  );
}
