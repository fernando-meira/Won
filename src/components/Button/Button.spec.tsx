import { screen } from '@testing-library/react';
import { renderWithTheme } from '~/utils/tests/helpers';
import { AddShoppingCart } from '@styled-icons/material/AddShoppingCart';

import Button from '.';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Buy now</Button>);

    expect(
      screen.getByRole('button', {
        name: /buy now/i,
      })
    ).toHaveStyle({
      height: '4rem',
      'font-size': '1.4rem',
      padding: '0.8rem 3.2rem',
    });

    expect(container).toMatchSnapshot();
  });

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Buy now</Button>);

    expect(
      screen.getByRole('button', {
        name: /buy now/i,
      })
    ).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem',
    });
  });

  it('should be render the large size', () => {
    renderWithTheme(<Button size="large">Buy now</Button>);

    expect(
      screen.getByRole('button', {
        name: /buy now/i,
      })
    ).toHaveStyle({
      height: '5rem',
      'font-size': '1.6rem',
      padding: '0.8rem 4.8rem',
    });
  });

  it('should be render fullWidth if prop exist', () => {
    renderWithTheme(<Button isFullWidth>Buy now</Button>);

    expect(
      screen.getByRole('button', {
        name: /buy now/i,
      })
    ).toHaveStyle({
      width: '100%',
    });
  });

  it('should be render icon version', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>
    );

    expect(screen.getByText(/buy now/i)).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toHaveStyle({
      width: '1.5rem',
    });
  });

  it('should be render anchor element', () => {
    const { debug, container } = renderWithTheme(
      <Button as="a" href="/link">
        Anchor test
      </Button>
    );

    debug(container);

    expect(
      screen.getByRole('link', {
        name: /Anchor test/i,
      })
    ).toHaveAttribute('href', '/link');
  });
});
