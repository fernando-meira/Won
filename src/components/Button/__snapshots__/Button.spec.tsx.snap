// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  border: none;
  padding: 0.8rem;
  border-radius: 0.4rem;
  cursor: pointer;
  color: #FAFAFA;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<div>
  <button
    class="c0"
  >
    <span>
      Buy now
    </span>
  </button>
</div>
`;
